### ISSUE TITLE
figure distortion

### ISSUE THREAD

**Issue by CaseyMiller** [brucemiller/LaTeXML#2392]
Figures become distorted. The example shows spheres in pane a becoming distorted to ellipsoids. Figure files attached and the source is

\documentclass[11pt]{article}

% Graphics support
\usepackage{graphicx}
\usepackage{subcaption}

\begin{document}

\begin{figure*}[t]
   \centering
    \begin{subfigure}[b]{0.33\linewidth}
        \includegraphics[width=\linewidth]{BCC.png}
        \caption{}
    \end{subfigure}
    \begin{subfigure}[b]{0.35\linewidth}
        \includegraphics[width=\linewidth]{BCE.png}
        \caption{}
    \end{subfigure}
    \begin{subfigure}[b]{0.3\linewidth}
        \includegraphics[width=\linewidth]{PSF_25S.png}
        \caption{}
    \end{subfigure}
    \\
    \begin{subfigure}[b]{0.4\linewidth}
        \includegraphics[width=\linewidth]{PEF_25E.png}
        \caption{}
    \end{subfigure}
    \begin{subfigure}[b]{0.4\linewidth}
        \includegraphics[width=\linewidth]{PSF_1000S.png}
        \caption{}
    \end{subfigure}
    \caption{Figure example.} 
    \label{fig:ex2}
\end{figure*}

\end{document}
![BCC](https://github.com/user-attachments/assets/f8914b59-b8de-4439-9771-b113e5d1fd56)
![BCE](https://github.com/user-attachments/assets/ec0c02bd-d01b-47bb-8bd4-ab23b588bf07)
![PEF_25E](https://github.com/user-attachments/assets/4814b86b-cc70-4834-a95c-caf2b5238b5f)
![PSF_25S](https://github.com/user-attachments/assets/d9022869-5e16-4e4b-a746-5a83af63966b)
![PSF_1000S](https://github.com/user-attachments/assets/189220c6-9ef5-40ed-8565-afc4bb0964d5)

**Comment by CaseyMiller** [comment]
[figure.pdf](https://github.com/user-attachments/files/18149696/figure.pdf)
[figure.zip](https://github.com/user-attachments/files/18149707/figure.zip)

Included example, pdf produced, and html.

**Comment by dginev** [comment]
The attached HTML keeps the row of 3, row of 2 arrangement from the PDF, and marks each of the 5 panels with `ltx_img_square`, which I think is close to the original image aspect ratios.

If there is visible distortion from here, I suspect it really is up to adjusting the `LaTeXML.css` styling rules.

**Comment by CaseyMiller** [comment]
compare the pdf and the html provided, there is substantial distortion

**Comment by CaseyMiller** [comment]
Updated files included
[figure.zip](https://github.com/user-attachments/files/18150477/figure.zip)

**Comment by brucemiller** [comment]
Queasily distorted, indeed.  If recognizing "almost square" helps identify a layout strategy, that's fine, but we should never change the aspect ratio from what is implied by the `\includegraphics` parameters. It's already risky enough to change the size.

**Comment by dginev** [comment]
See attached how it looks with `ar5iv.0.8.0.min.css` swapped in as the only CSS asset.
That seems to match at least my PDF viewer. I am not sure if Bruce's last comment implies LaTeXML.css should move further away from the ar5iv styling or closer. I'll leave that to him.

```css
<link rel="stylesheet" href="https://ar5iv.labs.arxiv.org/assets/ar5iv.0.8.0.min.css" type="text/css">
```

The reason it looks better is that it overrides the absolute `img` width and height size values with
```
.ltx_img_square {
    width: 100%;
    height: auto;
    max-height: var(--main-width);
}
```

As ar5iv aims (and will continue to aim) for responsive sizes.
In the mainline LaTeXML.css variant, the subfigure has an active `max-width` constraint, but nothing modifies the image height, leading to the aspect ratio change.

<img width="400px" src="https://github.com/user-attachments/assets/4d34ebd0-d604-4aeb-a349-6fc2b7ac645b">

**Comment by brucemiller** [comment]
Right; But LaTeXML itself aims to be responsive, as well, if perhaps not as extreme. The problem seems to be that the CSS is not divided up between stock & arxiv consistently.
The html img has explicit width and height, and the aspect ratio is correct. But the LaTeXML CSS has `max-width: calc(0.30*60rem);`. As you noted, that changes the width, but not the height, so the aspect ratio is now wrong.
Your arxiv stylesheet recovers by resetting height to `auto`, but I think that only works because the underlying image has the correct aspect ratio. If both width & height had been adjusted by `\includegraphics`, changing the aspect ratio, then this would not have worked. Any css that tweaks min/max/width/height has to tweak *both* the width & height to preserve the *requested* aspect ratio, not (necessarily) the aspect ratio of the original image.

**Comment by dginev** [comment]
> I think that only works because the underlying image has the correct aspect ratio.

Yes. If the underlying image did not have the correct aspect ratio, that would be a post-processing bug with ImageMagick. So it can/should be taken for granted when styling responsive subfigures.

**Comment by dginev** [comment]
>  If both width & height had been adjusted by \includegraphics, changing the aspect ratio, then this would not have worked. 

No, it would still work, because the `ltx_img_square` is computed as late as possible, only when the final width and height have become known in post-processing.

**Comment by brucemiller** [comment]
> Yes. If the underlying image did not have the correct aspect ratio, that would be a post-processing bug with ImageMagick. So it can/should be taken for granted when styling responsive subfigures.
No, the image hasn't necessarily been changed at all, and is not a bug: the image may simply be copied to the destination, and the `img` tag gets the width & height that give the desired aspect ratio.

**Comment by dginev** [comment]
> > Yes. If the underlying image did not have the correct aspect ratio, that would be a post-processing bug with ImageMagick. So it can/should be taken for granted when styling responsive subfigures.
> 
> No, the image hasn't necessarily been changed at all, and is not a bug: the image may simply be copied to the destination, and the `img` tag gets the width & height that give the desired aspect ratio.

I would treat that as either an author mistake or a latexml bug personally, especially if the aspect ratios disagree.

Also, if the final width & height values today represent a "desired aspect ratio", then so do the `ltx_image_square`, `ltx_image_portrait`, `ltx_image_landscape` classes that are derived from them. They could be further refined into a grid model if there was additional need for precision.

**Comment by brucemiller** [comment]
There is no bug, or necessarily mistake, for the height & width attributes (in xml or html) representing the desired image size and aspect ratio, and possibly being different from the height & width of the image file itself. That's the design, and html's img has worked that way since forever. When the aspect ratio is near enough to 1, then having class `ltx_image_square` is presumably also not an error; similarly/presumably for the landscape & portrait.
The bug is changing the width and/or height (even via min or max or even auto) without preserving the desired aspect ratio (as determined from the `\includegraphics` options and image file, but *not* exclusively from the image file).

**Comment by dginev** [comment]
I suspect we are talking cross-purposes here. The absolute sizes in the height and width attributes are not responsive, and can not by themselves be a foundation for a responsive design of subfigures.

That height+width setup was used in combination with the HTML `<table>` element to lay out subfigures in a rigid grid, which would lead to horizontal scroll on devices with narrow screens. Some additional adaptation is needed - the new flexbox-subfigures design tries one experiment in that direction. But the final touch-ups from ar5iv are missing from LaTeXML.css (which is the CSS file that leads to the reported bug).

Ensuring that the actual image asset has the desired aspect ratio during post-processing allows for easier responsive styling at the CSS stage. LaTeXML is at liberty to post-process images whenever it sees a need to, or not.

**Comment by brucemiller** [comment]
Of course explicit width & height aren't responsive. But then, neither is randomly changing the aspect ratio. I don't know where tables enter into this. And, sure, responsiveness is good, if done correctly.
Images are intentionally *not* rescaled in postprocessing, if it can be avoided. This is to avoid image duplication, extra passes through ImageMagick with all the time & bugs that encounters, and potential image degradation if the new resolution is too low (which also affects responsiveness). That decision is not cast in stone, but if it's changed, all those issues have to be resolved.   But assuming that the image has been rescaled just because that's what's easy in CSS isn't particularly productive.
Assigning to me is probably also counterproductive, since I'd just disable the feature until it can be done correctly.

**Comment by dginev** [comment]
Yes, this has been counter-productive for some time now.

> Assigning to me is probably also counterproductive, since I'd just disable the feature until it can be done correctly.

Sorry, as opposed to which other resolution? You seem to mostly ignore the various upgrade paths I outlined above. 

I can't be assigned to an issue without having a path to work on it by an objecting project owner. The absolute sizes in width and height can not be a foundation for a responsive design, and I want to work on a responsive design. Let me know if that interests you again at some future point.

**Comment by xworld21** [comment]
Since this issue seems to be mostly about `max-width: calc(0.30*60rem)`, I should add that it causes other issues. I had to disable it altogether in BookML because images containing text would become unreadable in certain cases. Readability should be the first concern, above responsiveness, and I don't think CSS resizing can be done safely with that in mind. (For reference, BookML rescales all SVG images so that the text inside roughly matches the font of the web page, which I think is the proper way to go here. That would include e.g. scaling up by 1.2x for documents at 10pt since browsers default at 12pt/16px fonts.)

**Comment by dginev** [comment]
>  Readability should be the first concern, above responsiveness, 

Great, I agree. But why not both?

> and I don't think CSS resizing can be done safely with that in mind. 

I think that should be tried in earnest, before the conclusion is reached with certainty. I would definitely appreciate an example (or several) of images causing problems, to check if there are possible improvements - as was the case in this issue.

**Comment by brucemiller** [comment]
As opposed to doing it correctly?  I don't know what a "foundation" for responsive design means, but it seems to me that having the absolute width & height implied by the author's markup, and thus the author's desired aspect ratio, is certainly the starting point for anything you'd want to do.  Conversely, ignoring the desired aspect ratio is doomed.
I didn't really see any  "upgrade paths" in the discussion above, other than having postprocessing only produce images with the target aspect ratio. That certainly is possible, but it is definitely a design *change* with a cost. Or alternatively, the CSS is more complicated.

**Comment by xworld21** [comment]
> > Readability should be the first concern, above responsiveness,
> 
> Great, I agree. But why not both?

After working on this for a while, I formed the conclusion that images should match their physical size in the PDF when possible, because that's the target that seems most likely to match the author's intention. Once you have images containing text, you can't really predict from LaTeXML if it's fine to change the font size or not. That's where I am coming from, at least (and I have the privilege of mostly handling scalable images, so I can tweak sizes for free).

In the example reported in this issue, where there is no text, `height: auto` is a good quick fix, though. At least the aspect ratio is correct. Certainly better than the current behaviour.

**Comment by dginev** [comment]
> I don't know what a "foundation" for responsive design means

We've been writing that word a lot recently - "ratio". That's what it means to me, relative sizes, made relative at the right time and towards some meaningful container. The current subfigure design does that via CSS, basing the aspect ratio on a rather "rough estimate" class attribute.

Then the main figure can be thought of as 100% of the width to fill, and each subfigure cell takes a fraction of that width, re-scaling its contents to fit as possible (allowing both width and height to adapt, to try and preserve the aspect ratio of the underlying image file. At least until there is some better, more precise, aspect ratio indication during the browser rendering stage).

> but it seems to me that having the absolute width & height implied by the author's markup, and thus the author's desired aspect ratio, is certainly the starting point for anything you'd want to do.

Yes, so much so that they are already used in computing the `ltx_image_square`, `ltx_image_portrait` and `ltx_image_landscape` classes. An upgrade path I mentioned (and was ignored) was to refine these classes further into some systematic classification of useful ratios. `ltx_image_16w_9h`, `ltx_image_1w_1h`, `ltx_image_4w_3h` ... No need to reinvent the wheel I suppos, [Tailwind CSS](https://tailwindcss.com/docs/aspect-ratio#arbitrary-values) has a technique for this using a class name `aspect-[4/3]`, as one example. Such classes can be emitted at the end of post-processing, similarly to how we emit `ltx_image_square` today.

And, as Vincenzo mentions, the simplest upgrade path was - and still is - carrying over the definitions from ar5iv.css to LaTeXML.css, which if nothing else is very quick to accomplish.

**Comment by brucemiller** [comment]
If I ignored the suggestion about `ltx_image_16w_9h` it is because I didn't see it. And after searching for it, I still don't see it.  I'm afraid to guess how adding that class would help; Please don't tell me that you want to force an image to have 16:9 aspect ratio just because the requested width & height were "near" that ratio? Surely if you're going to change the image size at all, you should preserve the requested aspect ratio exactly. (period).

I can see how your arrangement of subfigures could be attractive, "responsive" even, under the right circumstances. But under the wrong circumstances (eg. text in the figures) the result is quite bad.  For me, the most important part of the responsiveness work that was done on subfigures was the reflowing.  The resizing is not so essential.

And finally, the `height:auto` patch is just a short-term hack to close the very specific issue opened here; it doesn't solve the underlying issue, because it preserves the image file's aspect ratio, not the requested aspect ratio.

**Comment by dginev** [comment]
>  I still don't see it.

It wasn't expanded on. "[They could be further refined into a grid model if there was additional need for precision.](https://github.com/brucemiller/LaTeXML/issues/2392#issuecomment-2546013346)". 
My mental model there continues to be the [Bootstrap approach](https://getbootstrap.com/docs/4.0/layout/grid/) to responsive design. Exposing the aspect ratio is the first step, but how that is used depends on the host CSS theme for the web resource.

The questions the CSS could try to answer are "how should a landscape (or more precise, say 16-by-9) image be resized inside a flex row of 1 cell? 2 cells? 3 cells? etc." 
The experimentation with ar5iv showed that at least having a handle on the difference between square/landscape/portrait is needed to get the figure to grow/shrink appropriately in the available space.

The limitation I am unhappy with after working on this for some months, is that currently the flex subfigures are limited to equal subdivisions, which is often artificial. A landscape image adjacent to a portrait image are typically meant to be placed as panels with different widths, but the flex subfigure design today will treat them as the same `ltx_flex_size_2` cells and impose a max-width constraint of 50% the main width. Luckily that is very often a good guess, but far from always. (Note: the actual image width is not 50%, as the landscape/portrait classes allow for additional control, including whether an image should grow horizontally or vertically to fill the available space, and in which conditions).

So enhancing the flex cell sizing design is where my own intuition is headed to first for general improvements.

> Surely if you're going to change the image size at all, you should preserve the requested aspect ratio exactly. (period).

Ideally yes, assuming that is perfectly known. And `height:auto;` is a very useful way to do so -- but only if the aspect ratio of the actual image is the desired one. 

If this had to be done externally, a comment that comes to mind is that `style="aspect-ratio: 543 / 305;"` looks less useful to anchor further styling than a `class="ltx_img_aspect_16_9"` would be. But one could set *both* at the same time to address this.

> I can see how your arrangement of subfigures could be attractive, "responsive" even, under the right circumstances. But under the wrong circumstances (eg. text in the figures) the result is quite bad. 

Yes, but to me this statement extends to the general topic of "converting LaTeX to HTML". We pick our trade-offs. And not all of us have the same preferences, clearly.

**Comment by xworld21** [comment]
> If this had to be done externally, a comment that comes to mind is that `style="aspect-ratio: 543 / 305;"` looks less useful to anchor further styling than a `class="ltx_img_aspect_16_9"` would be. But one could set _both_ at the same time to address this.

Setting `style="aspect-ratio: @imagewidth / @imageheight"` in postprocessing should satisfy everybody here; it allows preserving the author's intent while allowing CSS resizing with `height: auto` (whether that happens in the default LaTeXML CSS or in further styling). The `16_9` class sounds to me like a hint for the flex styling, not the actual aspect ratio, since it can't be always accurate.

I still side with @brucemiller on this. If you don't resize images, the worst that can happen is that they reflow differently from PDF; if you resize them, the worst that can happen is that you can't read them. I find the latter much worse than the former.

**Comment by dginev** [comment]
@xworld21 is Figure 3 readable here to you? (no resizing at all):
https://ar5iv.labs.arxiv.org/html/2007.05629#S0.F3.g1

Switch that to `width: 100%; height: auto;` and compare.

A point I would like to re-raise is that we are solving different problems, and have different data to back up our design assumptions.

**Comment by brucemiller** [comment]
Well, you didn't ask me :>, but it's unreadable as it is. The change you suggest makes it readable, but fuzzy. The resolution is too low, but I'd have to dig in to determine whether it's due to the original (& markup), or a processing bug.  I was disturbed that it didn't reflow, until I realized it was a single image :>

**Comment by brucemiller** [comment]
Curious: If I convert a trivial document with that figure, I get a 435x372 pixel png; smaller than I'd have chosen, but still readable. And if I fetch the article from arXiv, I get the same image sizes.  But, the ar5iv version has 241x206 pixels. How'd that happen?

**Comment by brucemiller** [comment]
Oooh, forgot about those. A naive reading of option names would have made me expect the image to get the same html size, but the underlying image file to have 1.8 *higher* resolution (in reserve, so to speak).  But it looks like the underlying ar5iv image came out the same as local, just displayed smaller; as if `magnify` did nothing?

**Comment by dginev** [comment]
Sigh, another bug, sent a PR. Maybe it is time to complete this discussion and create new issues for any remaining conversations.

I'll let Bruce decide what to do with LaTeXML.css, which is what is being used here.
